### PR TITLE
Implement SFT dataloader checkpointing

### PR DESCRIPTION
**Pull request by mikasenghaas**
> Don't be scared by locs - most of it is unit testing the stateful dataset and dataloader:)) the loc change outside of tests dir is actually super small

This PR makes the SFT dataset and dataloader stateful to allow for distributed checkpointing without any changes to the `CheckpointManager`.

This is done by enforcing any SFT dataset to implement the `Stateful` protocol, which means it has to expose methods `state_dict` and `load_state_dict`:
- We define the base class `StatefulIterableDataset`. It contains logic for setting up the data rank and world size based on the number of data workers (local processes) and data ranks (DP world size) and initialize the dataset state, which only contains a `step` (how many samples have been seen globally) and `epoch` (which epoch are we in)
- Both the `FakeDataset` and `SFTDataset` inherit from this class and define a `__iter__` function as before but update the state in place
- The `PackingDataset` which is used to wrap the regular dataset simply passes the dataset state on. This means we count a sample as "used" even if we truncate it
- We replace the regular `torch.utils.data.DataLoader` with `torchdata.stateful_dataloader.StatefulDataloader` ([docs](https://github.com/pytorch/data/tree/main/torchdata/stateful_dataloader#saving-custom-state-with-iterable-style-datasets)). Because the iterable dataset we pass to it is rank- and worker-aware state, so is the dataloader
- We add extensive unit tests to ensure the logic is correct in `tests/unit/train/sft/test_{dataset,dataloader}` and show example runs in the PR description

Misc changes:
- Renamed test directory `tests/unit/training` -> `tests/unit/train`
- Fix correctly counting number of samples when packing (before we were incrementing by `batch_size` on each step which is only correct for padding which we deprecated)
- Moved some unit tests from `tests/unit/train` to `tests/unit/train/sft` and `tests/unit/train/rl` if the test is specific to that module
- Add the `num_examples` keyword for running SFT on a subset of data, here used for unit testing
- Allow `FakeData` to generate `fixed` or `variable` length inputs and `random` or `increasing` `input_ids` (the latter means that the first sample will be a tensor full of 0s, then 1s, etc.) - which is nice for unit testing the dataloader checkpointing
- Use discriminated union for `FakeDataConfig` and `SFTDataConfig` because the fields 

TODO:
- [x] Add fake data toy example (1 rank)
- [x] Add real data toy example (1 rank)
- [x] Add real data toy example (2 ranks)
- [x] Show checkpoint resuming on actual run

## Fake Data Toy Example

We do four steps on fake data with (micro) batch size 1 where the inputs IDs are tensors full of 0, 1, 2, ... (increasing). In this setup the dataloader step equals the fake input IDs.

```bash
uv run sft @ configs/reverse_text/sft.toml --log.level debug --data.type fake --data.length fixed --data.input_ids increasing --data.batch-size 1 --data.micro-batch-size 1 --max-steps 4 --ckpt
```

Screenshot 2025-08-28 at 2 25 46 PM

```bash
uv run sft @ configs/reverse_text/sft.toml --log.level debug --data.type fake --data.length fixed --data.input_ids increasing --data.batch-size 1 --data.micro-batch-size 1 --max-steps 8 --ckpt.resume-step 4
```

Screenshot 2025-08-28 at 2 26 45 PM

## Real Data Toy Example (1 Rank)

We start a tiny SFT run on real, packed data on a single rank for 4 steps and checkpoint at the very end.

```bash
uv run sft @ configs/reverse_text/sft.toml --log.level debug --max-steps 4 --ckpt
```

Screenshot 2025-08-28 at 2 20 54 PM

```bash
uv run sft @ configs/reverse_text/sft.toml --log.level debug --max-steps 8 --ckpt.resume-step 4
```

Screenshot 2025-08-28 at 2 21 38 PM

As we can see, we  from checkpoint and dataloader step 0. With each step we increment the dataloader step by ~20  which makes sense given that in each step we see `batch_size (16) * seq_len (128) = 2048` tokens, which is roughly 2048/20≈102 tokens/per sample which is true for this dataset. Upon restart , we resume with step 4 and and dataloader step 89, as expected.

## Real Data Toy Example (2 Ranks)

We start a tiny SFT run on real, packed data on a two ranks rank for 4 steps and checkpoint at the very end.

```bash
uv run torchrun --nproc-per-node 2 src/prime_rl/trainer/sft/train.py @ configs/reverse_text/sft.toml --log.level debug --max-steps 4 --ckpt
```

Screenshot 2025-08-28 at 2 42 19 PM

```bash
uv run torchrun --nproc-per-node 2 src/prime_rl/trainer/sft/train.py @ configs/reverse_text/sft.toml --log.level debug --max-steps 8 --ckpt.resume-step 4
```

Screenshot 2025-08-28 at 2 41 47 PM

## Real Data Example

[W&B View](https://wandb.ai/primeintellect/mika?nw=z5gu3zpiq4p)

Screenshot 2025-08-28 at 2 46 46 PM

```bash
uv run torchrun --nproc-per-node 2 src/prime_rl/trainer/sft/train.py @ configs/reverse_text/sft.toml --log.level debug --monitor.wandb.project mika --monitor.wandb.name pr-849 --max-steps 50 --ckpt
```

```bash
uv run torchrun --nproc-per-node 2 src/prime_rl/trainer/sft/train.py @ configs/reverse_text/sft.toml --log.level debug --monitor.wandb.project mika --monitor.wandb.name pr-849-resume --max-steps 100 --ckpt.resume-step 50
```

---

**GitHub Issue**: #745 
**Linear Issue**: Resolves PRIMERL-19

### CODE REVIEW

**Comment by mikasenghaas** [comment]
Thanks @baonudesifeizhai for pioneering this!

File: src/prime_rl/trainer/sft/data.py [hunk 1]
```diff
@@ -29,41 +31,85 @@ class Batch(TypedDict):
...
+        # -1 because the stateful dataloader expects uses 1-based counting while we start at 0
+        self.step = state_dict["step"] - 1
+        self.epoch = state_dict["epoch"]
+
+    def _setup_world_info(self):
+        worker_info = get_worker_info()
+        worker_id, num_workers = 0, 1
+        if worker_info is not None:
+            worker_id = worker_info.id
+            num_workers = worker_info.num_workers
```
Comment (samsja) [inline src/prime_rl/trainer/sft/data.py]:
```suggestion
        if worker_info is not None:
            worker_id = worker_info.id
            num_workers = worker_info.num_workers
        else: 
              worker_id, num_workers = 0, 1

```

Comment (mikasenghaas) [reply src/prime_rl/trainer/sft/data.py]: spacing looked off here, committed separate

**Review by Jackmin801 — APPROVED** [review]
letsgoo

**Review by Jackmin801 — APPROVED** [review]
nice! test looks like it should test the correct behavior. ship it!
